File: things/src/main/java/com/krisbiketeam/smarthomeraspbpi3/driver/TMP102.java
package com.krisbiketeam.smarthomeraspbpi3.driver;

import android.support.annotation.IntDef;
import android.support.annotation.VisibleForTesting;

import com.google.android.things.pio.I2cDevice;
import com.google.android.things.pio.PeripheralManager;

import java.io.IOException;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import timber.log.Timber;

/**
 * Driver for the TMP102 temperature sensor.
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class TMP102 implements AutoCloseable {

    /**
     * Default I2C address for the sensor.
     */
    public static final int DEFAULT_I2C_GND_ADDRESS = 0x48;
    public static final int DEFAULT_I2C_VCC_ADDRESS = 0x49;
    public static final int DEFAULT_I2C_SDA_ADDRESS = 0x4A;
    public static final int DEFAULT_I2C_SCL_ADDRESS = 0x4B;
    @Deprecated
    public static final int I2C_ADDRESS = DEFAULT_I2C_GND_ADDRESS;

    // Sensor constants from the datasheet.
    // https://cdn-shop.adafruit.com/datasheets/BST-BMP280-DS001-11.pdf
    /**
     * Minimum temperature in Celsius the sensor can measure.
     */
    public static final float MIN_TEMP_C = -25f;
    /**
     * Maximum temperature in Celsius the sensor can measure.
     */
    public static final float MAX_TEMP_C = 85f;
    /**
     * Maximum power consumption in micro-amperes when measuring temperature.
     */
    public static final float MAX_POWER_CONSUMPTION_TEMP_UA = 85f;
    /**
     * Maximum frequency of the measurements.
     */
    public static final float MAX_FREQ_HZ = 8f;
    /**
     * Minimum frequency of the measurements.
     */
    public static final float MIN_FREQ_HZ = 1f;

    /**
     * Extended mode EM.
     */
    @Retention(RetentionPolicy.SOURCE)
    @IntDef({NOT_EXTENDED_MODE, EXTENDED_MODE,})
    public @interface ExtendedMode {
    }

    public static final int NOT_EXTENDED_MODE = 0;      // default
    public static final int EXTENDED_MODE = 1;

    private static final int TMP102_EXTENDED_MODE_BIT_SHIFT = 4;

    /**
     * Extended mode CR1 and CR0.
     * //0 1 CONVERSION_RATE4 is default
     */
    @Retention(RetentionPolicy.SOURCE)
    @IntDef({CONVERSION_RATE025, CONVERSION_RATE1, CONVERSION_RATE4, CONVERSION_RATE8,})
    public @interface ConversionRate {
    }

    public static final int CONVERSION_RATE025 = 0;
    public static final int CONVERSION_RATE1 = 1;
    public static final int CONVERSION_RATE4 = 2;
    public static final int CONVERSION_RATE8 = 3;

    private static final int TMP102_CONVERSION_RATE_BIT_SHIFT = 6;
    private static final int TMP102_CONVERSION_RATE_MASK = 0b11000000;

    /**
     * Shutdown mode SD.
     * The Shutdown mode bit saves maximum power by Figure 10. Output Transfer Function Diagrams
     * shutting down all device circuitry other than the serial interface, reducing current
     * consumption to typically less than 0.5mA. Shutdown mode is enabled when the SD bit is '1';
     * the device shuts down when current conversion is completed. When SD is equal to '0', the A
     * fault condition exists when the measured device maintains a continuous conversion state.
     */
    @Retention(RetentionPolicy.SOURCE)
    @IntDef({NOT_SHUTDOWN_MODE, SHUTDOWN_MODE,})
    public @interface ShutdownMode {
    }

    public static final int NOT_SHUTDOWN_MODE = 0;
    public static final int SHUTDOWN_MODE = 1;

    private static final int TMP102_SHUTDOWN_MODE_BIT_SHIFT = 8;


    // Registers
    private static final int TMP102_REG_TEMP = 0x00;
    @VisibleForTesting
    static final int TMP102_REG_CONF = 0x01;
    private static final int TMP102_REG_T_LOW = 0x02;
    private static final int TMP102_REG_T_HIGH = 0x03;

    private static final float TEMP_REG_FACTOR = 0.0625f;

    private static final int BMX280_OVERSAMPLING_PRESSURE_MASK = 0b00011100;
    private static final int BMX280_OVERSAMPLING_PRESSURE_BITSHIFT = 2;
    private static final int BMX280_OVERSAMPLING_TEMPERATURE_MASK = 0b11100000;
    private static final int BMX280_OVERSAMPLING_TEMPERATURE_BITSHIFT = 5;

    private I2cDevice mDevice;
    private final byte[] mBuffer = new byte[2]; // for reading sensor values
    private boolean mEnabled;
    private int mConfig;

    /**
     * Create a new TMP102 sensor driver connected on the given bus.
     *
     * @param bus I2C bus the sensor is connected to.
     * @throws IOException
     */
    public TMP102(String bus) throws IOException {
        this(bus, DEFAULT_I2C_GND_ADDRESS);
    }

    /**
     * Create a new TMP102 sensor driver connected on the given bus and address.
     *
     * @param bus     I2C bus the sensor is connected to.
     * @param address I2C address of the sensor.
     * @throws IOException
     */
    public TMP102(String bus, int address) throws IOException {
        PeripheralManager pioService = PeripheralManager.getInstance();
        I2cDevice device = pioService.openI2cDevice(bus, address);
        try {
            connect(device);
        } catch (IOException | RuntimeException e) {
            try {
                close();
            } catch (IOException | RuntimeException ignored) {
            }
            throw e;
        }
    }

    /**
     * Create a new TMP102 sensor driver connected to the given I2c device.
     *
     * @param device I2C device of the sensor.
     * @throws IOException
     */
    /*package*/  TMP102(I2cDevice device) throws IOException {
        connect(device);
    }

    private void connect(I2cDevice device) throws IOException {
        mDevice = device;
        mConfig = readSample16(TMP102_REG_CONF);
        Timber.d("connect mConfig: " + mConfig);
    }

    /**
     * Close the driver and the underlying device.
     */
    @Override
    public void close() throws IOException {
        if (mDevice != null) {
            try {
                mDevice.close();
            } finally {
                mDevice = null;
            }
        }
    }

    /**
     * Check if TMP102 is in Extended Mode where 13-bits of Temperature register are read
     *
     * @return true if sensor is in EM (Extended Mode) mode
     * @throws IOException
     */
    public boolean isExtendedMode() throws IOException {
        // is this really needed?
        mConfig = readSample16(TMP102_REG_CONF);
        return isExtendedMode(mConfig);
    }

    @VisibleForTesting
    boolean isExtendedMode(int config) {
        return (1 << TMP102_EXTENDED_MODE_BIT_SHIFT & config) > 0;
    }

    /**
     * Set TMP102  in Extended Mode where 13-bits of Temperature register are read.
     *
     * @param extended true if we want to read 13 bit temperature register (Extended Mode)
     * @throws IOException
     */
    public void setExtendedMode(boolean extended) throws IOException {
        // is this really needed?
        mConfig = readSample16(TMP102_REG_CONF);
        setExtendedMode(mConfig, extended);
    }

    @VisibleForTesting
    int setExtendedMode(int config, boolean extended) throws IOException {
        if (extended) {
            config |= EXTENDED_MODE << TMP102_EXTENDED_MODE_BIT_SHIFT;
        } else {
            config &= ~(EXTENDED_MODE << TMP102_EXTENDED_MODE_BIT_SHIFT);
        }
        writeSample16(TMP102_REG_CONF, config);
        return config;
    }

    /**
     * Set the power mode of the sensor.
     *
     * @param mode power mode.
     * @throws IOException
     */
    public void setConversionRateMode(@ConversionRate int mode) throws IOException {
        if (mDevice == null) {
            throw new IllegalStateException("I2C device not open");
        }

        mConfig = readSample16(TMP102_REG_CONF);
        switch (mode) {
            case CONVERSION_RATE025:
            case CONVERSION_RATE1:
            case CONVERSION_RATE4:
            case CONVERSION_RATE8:
                mConfig &= ~TMP102_CONVERSION_RATE_MASK;
                mConfig |= mode << TMP102_CONVERSION_RATE_BIT_SHIFT;
                break;
            default:
                Timber.w("setConversionRateMode wrong Conversion mode");
        }
        writeSample16(TMP102_REG_CONF, mConfig);
    }

    /**
     * Read the current temperature.
     *
     * @return the current temperature in degrees Celsius
     */
    public float readTemperature() throws IOException {
        int rawTemp = readSample16(TMP102_REG_TEMP);
        return calculateTemperature(rawTemp);
    }

    /**
     * Reads 16 bits from the given address.
     *
     * @throws IOException
     */
    private int readSample16(int address) throws IOException, IllegalStateException {
        if (mDevice == null) {
            throw new IllegalStateException("I2C device is already closed");
        }
        synchronized (mBuffer) {
            // Reading a byte buffer instead of a short to avoid having to deal with
            // platform-specific endianness.
            mDevice.readRegBuffer(address, mBuffer, 2);
            // msb[7:0] lsb[7:0]

            int msb = mBuffer[0] & 0xff;
            int lsb = mBuffer[1] & 0xff;
            return msb << 8 | lsb;
        }
    }

    private void writeSample16(int address, int data) throws IOException, IllegalStateException {
        if (mDevice == null) {
            throw new IllegalStateException("I2C device is already closed");
        }
        synchronized (mBuffer) {
            //msb
            mBuffer[0] = (byte) (data >> 8);
            mBuffer[1] = (byte) (data);
            // Reading a byte buffer instead of a short to avoid having to deal with
            // platform-specific endianness.
            mDevice.writeRegBuffer(address, mBuffer, 2);
        }
    }

    /**
     * Calculate real temperature in Celsius degree from Raw temp value
     *
     * @param rawTemp Raw temperature returned from TMP102 Sensor
     * @return
     */
    @VisibleForTesting
    float calculateTemperature(int rawTemp) {
        if (rawTemp < 0x8000) {
            // Check if raw Data is not in extended EM 13 bit mode
            if ((rawTemp & 0x01) > 0) {
                return (rawTemp >> 3) * TEMP_REG_FACTOR;
            } else {
                return (rawTemp >> 4) * TEMP_REG_FACTOR;
            }
        } else {    // Negative number of 2's compliment
            // Check if raw Data is not in extended EM 13 bit mode
            if ((rawTemp & 0x01f) > 0) {
                rawTemp = ~(rawTemp - 0b1000);
                rawTemp = (rawTemp & 0xFFFF);
                return (rawTemp >> 3) * -TEMP_REG_FACTOR;
            } else {
                rawTemp = ~(rawTemp - 0b10000);
                rawTemp = (rawTemp & 0xFFFF);
                return (rawTemp >> 4) * -TEMP_REG_FACTOR;
            }
        }
    }
}

File: things/src/main/java/com/krisbiketeam/smarthomeraspbpi3/driver/TMP102.kt
package com.krisbiketeam.smarthomeraspbpi3.driver

import android.support.annotation.VisibleForTesting
import com.google.android.things.pio.I2cDevice
import com.google.android.things.pio.PeripheralManager
import kotlinx.coroutines.experimental.delay
import kotlinx.coroutines.experimental.launch
import timber.log.Timber
import java.io.IOException

/**
 * Driver for the TMP102 temperature sensor.
 */
class TMP102(bus: String? = null, address: Int = DEFAULT_I2C_GND_ADDRESS) : AutoCloseable {


    companion object {

        /**
         * Default I2C address for the sensor.
         */
        const val DEFAULT_I2C_GND_ADDRESS = 0x48
        const val DEFAULT_I2C_VCC_ADDRESS = 0x49
        const val DEFAULT_I2C_SDA_ADDRESS = 0x4A
        const val DEFAULT_I2C_SCL_ADDRESS = 0x4B

        @Deprecated("")
        const val I2C_ADDRESS = DEFAULT_I2C_GND_ADDRESS

        // Sensor constants from the datasheet.
        // https://cdn-shop.adafruit.com/datasheets/BST-BMP280-DS001-11.pdf
        /**
         * Minimum temperature in Celsius the sensor can measure.
         */
        const val MIN_TEMP_C = -25f
        /**
         * Maximum temperature in Celsius the sensor can measure.
         */
        const val MAX_TEMP_C = 85f
        /**
         * Maximum power consumption in micro-amperes when measuring temperature.
         */
        const val MAX_POWER_CONSUMPTION_TEMP_UA = 10f
        /**
         * Maximum frequency of the measurements.
         */
        const val MAX_FREQ_HZ = 8f
        /**
         * Minimum frequency of the measurements.
         */
        const val MIN_FREQ_HZ = 0.25f

        /**
         * Maximum frequency of the measurements.
         */
        const val POWER_ON_CONVERSION_DELAY = 26L


        private const val TMP102_EXTENDED_MODE_BIT_SHIFT = 4

        private const val TMP102_CONVERSION_RATE_BIT_SHIFT = 6
        private const val TMP102_CONVERSION_RATE_MASK = 192

        private const val TMP102_SHUTDOWN_MODE_BIT_SHIFT = 8
        private const val TMP102_ONE_SHOT_BIT_SHIFT = 15

        // Registers
        private const val TMP102_REG_TEMP = 0x00
        @VisibleForTesting
        internal const val TMP102_REG_CONF = 0x01
        private const val TMP102_REG_T_LOW = 0x02
        private const val TMP102_REG_T_HIGH = 0x03

        private const val TEMP_REG_FACTOR = 0.0625f
    }

    private var mDevice: I2cDevice? = null
    private val mBuffer = ByteArray(2) // for reading sensor values
    @VisibleForTesting
    internal var mConfig: Int = 0

    /**
     * Extended mode EM.
     */
    enum class ExtendedMode(val value: Int) {
        NOT_EXTENDED_MODE(0),      // default
        EXTENDED_MODE(1);
    }

    var extendedMode: Boolean
        /**
         * Check if TMP102 is in Extended Mode where 13-bits of Temperature register are read
         *
         * @return true if sensor is in EM (Extended Mode) mode
         */
        get() {
            return 1 shl TMP102_EXTENDED_MODE_BIT_SHIFT and mConfig > 0
        }
        /**
         * Set TMP102  in Extended Mode where 13-bits of Temperature register are read.
         *
         * @param extended true if we want to read 13 bit temperature register (Extended Mode)
         * @throws IOException
         */
        set(extended) {
            if (extended) {
                mConfig = mConfig or (1 shl TMP102_EXTENDED_MODE_BIT_SHIFT)
            } else {
                mConfig = mConfig and (1 shl TMP102_EXTENDED_MODE_BIT_SHIFT).inv()
            }
            writeSample16(TMP102_REG_CONF, mConfig)
        }

    /**
     * Conversion mode CR1 and CR0.
     * "01" CONVERSION_RATE4 is default
     */
    enum class ConversionRate(val value: Int) {
        CONVERSION_RATE025(0),
        CONVERSION_RATE1(1),
        CONVERSION_RATE4(2),
        CONVERSION_RATE8(3);
    }

    var conversionRateMode: ConversionRate
        /**
         * Get the power mode of the sensor.
         *
         * @return power mode.
         */
        get() {
            var tmp = mConfig and TMP102_CONVERSION_RATE_MASK
            tmp = (tmp shr TMP102_CONVERSION_RATE_BIT_SHIFT)
            when (tmp) {
                3 -> return ConversionRate.CONVERSION_RATE8
                2 -> return ConversionRate.CONVERSION_RATE4
                1 -> return ConversionRate.CONVERSION_RATE1
                else -> return ConversionRate.CONVERSION_RATE025
            }
        }
        /**
         * Set the power mode of the sensor.
         *
         * @param mode power mode.
         */
        set(mode) {
            mConfig = mConfig and TMP102_CONVERSION_RATE_MASK.inv()
            mConfig = mConfig or (mode.value shl TMP102_CONVERSION_RATE_BIT_SHIFT)
            writeSample16(TMP102_REG_CONF, mConfig)
        }

    /**
     * Shutdown mode SD.
     * The Shutdown mode bit saves maximum power by Figure 10. Output Transfer Function Diagrams
     * shutting down all device circuitry other than the serial interface, reducing current
     * consumption to typically less than 0.5mA. Shutdown mode is enabled when the SD bit is '1';
     * the device shuts down when current conversion is completed. When SD is equal to '0', the A
     * fault condition exists when the measured device maintains a continuous conversion state.
     */
    enum class ShutdownMode(val value: Int) {
        CONTINOUS_MODE(0),
        SHUTDOWN_MODE(1)
    }

    var shutdownMode: Boolean
        /**
         * Check if TMP102 is in shutdown mode where device gets into sleep mode after temp read
         *
         * @return true if sensor is in SD (ShutDown Mode) mode
         */
        get() {
            return 1 shl TMP102_SHUTDOWN_MODE_BIT_SHIFT and mConfig > 0
        }
        /**
         * Set TMP102 in ShutDown Mode where device gets into sleep mode after temp read
         *
         * @param extended true if we want to switch to SD (ShutDown) mode
         * @throws IOException
         */
        set(shutdown) {
            if (shutdown) {
                mConfig = mConfig or (1 shl TMP102_SHUTDOWN_MODE_BIT_SHIFT)
            } else {
                mConfig = mConfig and (1 shl TMP102_SHUTDOWN_MODE_BIT_SHIFT).inv()
            }
            writeSample16(TMP102_REG_CONF, mConfig)
        }


    init {
        if (bus != null) {
            val device = PeripheralManager.getInstance().openI2cDevice(bus, address)
            try {
                mDevice = device
                mConfig = readSample16(TMP102_REG_CONF) ?: 0
                Timber.d("connect mConfig: $mConfig")
            } catch (e: IOException) {
                Timber.e("Error connecting device", e)
                try {
                    close()
                } catch (e: IOException) {
                    Timber.e("Error closing device", e)
                }
            }
        }
    }

    /**
     * Create a new TMP102 sensor driver connected to the given I2c device.
     *
     * @param device I2C device of the sensor.
     * @throws IOException
     */
    @VisibleForTesting
    internal constructor(device: I2cDevice, config: Int) : this() {
        mDevice = device
        mConfig = config
    }

    /**
     * Close the driver and the underlying device.
     */
    override fun close() {
        try {
            mDevice?.close()
        } finally {
            mDevice = null
        }
    }


    /**
     * Read the current temperature.
     *
     * @return the current temperature in degrees Celsius
     */
    fun readTemperature(): Float? = calculateTemperature(readSample16(TMP102_REG_TEMP))

    /**
     * Read the current temperature in SD (ShutDown) mode. Callback will be triggered after temp read is completed
      */
    fun readOneShotTemperature(onResult: (Float?)-> Unit) {
        if (shutdownMode) {
            synchronized(mBuffer) {
                launch {
                    // Write OneShot bit to config to wakeupd device for one shot read temp
                    mConfig = mConfig or (1 shl TMP102_ONE_SHOT_BIT_SHIFT)
                    writeSample16(TMP102_REG_CONF, mConfig)
                    // Wait 26 ms for conversion to complete
                    delay(POWER_ON_CONVERSION_DELAY)
                    // check if conversion finished by reading OS bit to '1'
                    if (1.shl(TMP102_ONE_SHOT_BIT_SHIFT).and(readSample16(TMP102_REG_CONF) ?: 0) > 0) {
                        val temp = readTemperature()
                        Timber.d("readOneShotTemperature conversion finished temp? $temp")
                        onResult(temp)
                    } else {
                        Timber.d("readOneShotTemperature conversion did not completed")
                        onResult(null)
                    }
                }
            }
        } else {
            val temp = readTemperature()
            onResult(temp)
        }
    }


    /**
     * Reads 16 bits from the given address.
     */
    private fun readSample16(address: Int): Int? {
        synchronized(mBuffer) {
            // Reading a byte buffer instead of a short to avoid having to deal with
            // platform-specific endianness.
            try {
                mDevice?.readRegBuffer(address, mBuffer, 2) ?: return null
            } catch (e: IOException) {
                Timber.e("Error reading RegBuffer", e)
                return null
            }
            // msb[7:0] lsb[7:0]

            val msb = mBuffer[0].toInt().and(0xff)
            val lsb = mBuffer[1].toInt().and(0xff)
            return msb shl 8 or lsb
        }
    }

    private fun writeSample16(address: Int, data: Int) {
        synchronized(mBuffer) {
            //msb
            mBuffer[0] = (data shr 8).toByte()
            mBuffer[1] = data.toByte()
            // Reading a byte buffer instead of a short to avoid having to deal with
            // platform-specific endianness.
            try {
                mDevice?.writeRegBuffer(address, mBuffer, 2)
            } catch (e: IOException) {
                Timber.e("Error writing RegBuffer", e)
            }
        }
    }

    /**
     * Calculate real temperature in Celsius degree from Raw temp value
     *
     * @param rawTemp Raw temperature returned from TMP102 Sensor
     * @return
     */
    @VisibleForTesting
    internal fun calculateTemperature(rawTemp: Int?): Float? {
        if (rawTemp == null) return null
        if (rawTemp < 0x8000) {
            // Check if raw Data is not in extended EM 13 bit mode
            return if (rawTemp and 0x01 > 0) {
                (rawTemp shr 3) * TEMP_REG_FACTOR
            } else {
                (rawTemp shr 4) * TEMP_REG_FACTOR
            }
        } else {    // Negative number of 2's compliment
            // Check if raw Data is not in extended EM 13 bit mode
            if (rawTemp and 0x01f > 0) {
                return ((((rawTemp - 8).inv()) and 0xFFFF) shr 3) * -TEMP_REG_FACTOR
            } else {
                return ((((rawTemp - 16).inv()) and 0xFFFF) shr 4) * -TEMP_REG_FACTOR
            }
        }
    }
}

